docs/upgrading_to_nameko_3.rst

```diff
@@ -0,0 +1,347 @@
+            # ...
+
+In Nameko 3, the above will result in a `DeprecationWarning`:
+
+.. code-block:
+
+    .../nameko/containers.py:173: DeprecationWarning: Use ``nameko.config`` instead.
+    warnings.warn("Use ``nameko.config`` instead.", DeprecationWarning)
+
+
+Programmatically running services
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning:: Breaking Change!
+
+    The following section contains a breaking change.
+
+This section is relevant if you are creating runner programmatically, rather than using the `run` CLI command.
+
+In Nameko 3, the `ServiceContainer` and `ServiceRunner` classes do not accept the `config` argument anymore. You must update configuration in the global object, with `config.patch` or directly:
+
+.. code-block:: python
+
+    import nameko
+    from nameko.containers import ServiceContainer
+
+    nameko.config["AMQP_URI"] = "pyamqp://someuser:*****@somehost/"
+
+    container = ServiceContainer(Service)
+    container.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    container.stop()
+
+Same for the runner:
+
+.. code-block:: python
+
+    from nameko.runners import ServiceRunner
+
+    nameko.config["AMQP_URI"] = "pyamqp://someuser:*****@somehost/"
+
+    runner = ServiceRunner()
+    runner.add_service(ServiceA)
+    runner.add_service(ServiceB)
+
+    runner.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    runner.stop()
+
+Other configuration related changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The `--broker`` CLI option is deprecated in favour of `--define` and `--config`. TODO: doesn’t actually raise a warning anymore
+* The built-in Config dependency provider is deprecated as the config can be accessed and read directly.
+
+Testing services
+----------------
+
+In Nameko 2, configuration was always explicitly passed as an object, which made it easy to manage in tests. The global config object needs some different ways of managing configuration in tests.
+
+Config patching
+~~~~~~~~~~~~~~~
+
+The global `config` object has a `patch` helper object which can be used as a context manager or as a decorator. Here is an example of patching the config object in a pytest fixture:
+
+.. code-block:: python
+
+    import nameko
+    import pytest
+
+    @pytest.fixture
+    def memory_rabbit_config():
+        with nameko.config.patch({"AMQP_URI": "memory://"}):
+            yield
+
+While this fixture is in place, any code accessing `config['AMQP_URI']` will receive the value `"memory://"`.
+
+Changes to the `container_factory` and `runner_factory` fixtures.
+
+In Nameko 2, the `container_factory` and `runner_factory` fixtures accepted `config` as a required argument. In Nameko 3, it is optional. If passed, `nameko.config.patch` is used to wrap running container. The patch is applied before container start and returned it back to its original value when exiting the test:
```

Comment:
Reformulate:

> [...]. If `config` is passed, the factory will use `config.patch()` to emulate the old behaviour. This will cause the config values to be reverted to their original values once the test has exited.